core/loader_shared.c

```diff
@@ -163,6 +165,7 @@ loader_should_call_entry(privmod_t *mod)
 void
 loader_init_prologue(void)
 {
+#ifdef WINDOWS
```

Comment:
I'm having a hard time understanding the changes from just the code: please augment the PR description to explain what the changes are.  It looks like we will now have even more divergent ordering across the platforms which will make it harder to maintain down the road.  In HEAD, I see just one difference, which is clearly explained in a comment: imports+TLS either before or after thread init.  With this PR it is hard to understand all the ordering differences.  I thought there were other solutions proposed and tried in #4501 which left the code in a cleaner state?

Reply:
This partial reversion was a last resort after I couldn't find any other way of making it work. I've added to the log message with a link to an issue to unify the Unix and Windows loaders.

Reply:
If you spent time on the other proposed #4501 solutions, please update the issue: there is no reply to https://github.com/DynamoRIO/dynamorio/issues/4501#issuecomment-1779724714 .  You tried what @yury-khrustalev reported to have worked at https://github.com/DynamoRIO/dynamorio/issues/4501#issuecomment-1105184395?  If the fundamental issue is that the module list has not-yet-initialized modules on it yet those uninit modules cannot possibly supply imports to a module whose imports are processed early, then just skipping when pd==NULL may be a simple solution if it is difficult to initialize the pd earlier (see discussion on another proposed solution https://github.com/DynamoRIO/dynamorio/issues/4501#issuecomment-1105284385) or if we did it would just be empty anyway.

I'm just seeing things in PR #4035 like "Reverses the modlist iteration order for function calling, to properly call independent libraries first before their dependents" which don't sound like things we want to revert: trying to avoid breaking something else by going backward.  Does the skip-lookup-if-pd==NULL solve this?

Reply:
I did try what @yury-khrustalev suggested, and it didn't work, as I recall, but I tried it again just now and it did seem to work. So unless my memory is at fault, either I made a silly mistake back then, or I made a silly mistake just now, or the difference is because of something else changing, which doesn't seem very likely. I'll check it again and perhaps I'll have a much smaller PR on 10 Sep 2024.

Reply:
It's a separate PR as the patch is totally different: #6976